File: quasar-backend/src/controllers/jdController.js
const { extractTextFromPDF } = require('../utils/pdfExtract');
const { chatCompletion } = require('../services/groqService');
const JdSession = require('../models/JdSession');
const JdQuestion = require('../models/JdQuestion');
const logger = require('../utils/logger');

/**
 * POST /api/jd/parse
 * Parse a job description and generate custom question bank (powered by Groq)
 */
async function parseJD(req, res) {
  try {
    const userId = req.user?.id;

    // Support both PDF upload (multipart) and plain-text body
    let jobDescription = req.body?.jobDescription?.trim() || '';

    if (req.file) {
      try {
        jobDescription = await extractTextFromPDF(req.file.buffer);
      } catch (pdfErr) {
        logger.error('PDF extraction failed on JD upload', { err: pdfErr.message });
        return res.status(422).json({ success: false, message: `Could not extract text from JD PDF: ${pdfErr.message}`, data: null });
      }
    }

    if (!jobDescription || jobDescription.length < 50) {
      return res.status(400).json({ success: false, message: 'Job description must be at least 50 characters', data: null });
    }

    const systemPrompt = `You are an expert technical recruiter. Analyse the provided job description and return ONLY valid JSON (no markdown, no code fences) matching this exact schema:
{
  "role": string,
  "seniority": "junior" | "mid" | "senior" | "staff" | "principal",
  "domain": string,
  "requiredSkills": string[],
  "niceToHaveSkills": string[],
  "culturalSignals": string[],
  "generatedQuestions": [
    {
      "question": string,
      "category": "behavioural" | "technical" | "system-design" | "culture-fit",
      "difficulty": 1 | 2 | 3,
      "targetSkill": string,
      "weight": number
    }
  ]
}
Generate exactly 20 questions, weighted by importance to the role. Weight values should be between 0.0 and 1.0, higher = more important.`;

    const userPrompt = `Job Description:\n${jobDescription}`;

    const responseText = await chatCompletion(systemPrompt, userPrompt, {
      model: 'llama-3.3-70b-versatile',
      temperature: 0.4,
      maxTokens: 4096,
    });

    // Parse JSON from response — strip markdown fences if present
    let cleanJson = responseText.trim();
    if (cleanJson.startsWith('```')) {
      cleanJson = cleanJson.replace(/^```(?:json)?\s*/i, '').replace(/```\s*$/, '');
    }

    let parsedData;
    try {
      parsedData = JSON.parse(cleanJson);
    } catch (parseErr) {
      logger.error('Failed to parse Groq JD response as JSON', { responseText, err: parseErr.message });
      return res.status(502).json({ success: false, message: 'AI returned invalid JSON. Please try again.', data: null });
    }

    // Store JD session
    const jdSession = await JdSession.create({
      userId,
      jobDescription,
      parsedData,
    });

    // Store individual questions
    if (parsedData.generatedQuestions && Array.isArray(parsedData.generatedQuestions)) {
      const questions = parsedData.generatedQuestions.map(q => ({
        jdSessionId: jdSession._id,
        question: q.question,
        category: q.category,
        difficulty: q.difficulty,
        targetSkill: q.targetSkill,
        weight: q.weight,
      }));
      await JdQuestion.insertMany(questions);
    }

    return res.json({
      success: true,
      message: 'Job description parsed successfully',
      data: {
        jdSessionId: jdSession._id,
        role: parsedData.role,
        seniority: parsedData.seniority,
        domain: parsedData.domain,
        requiredSkills: parsedData.requiredSkills,
        niceToHaveSkills: parsedData.niceToHaveSkills,
        culturalSignals: parsedData.culturalSignals,
        questions: parsedData.generatedQuestions
      }
    });
  } catch (err) {
    logger.error('Parse JD error', { err: err.message });
    return res.status(500).json({ success: false, message: 'Failed to parse job description', data: null });
  }
}

/**
 * GET /api/jd/:jdSessionId/questions
 * Returns question list ordered by weight desc
 */
async function getJDQuestions(req, res) {
  try {
    const { jdSessionId } = req.params;

    const questions = await JdQuestion.find({ jdSessionId }).sort({ weight: -1 }).lean();

    return res.json({
      success: true,
      message: 'Questions retrieved',
      data: questions
    });
  } catch (err) {
    logger.error('Get JD questions error', { err: err.message });
    return res.status(500).json({ success: false, message: 'Failed to get questions', data: null });
  }
}

module.exports = { parseJD, getJDQuestions };

File: backend/src/main/java/com/teamzemo/quasar/controller/JdController.java
package com.teamzemo.quasar.controller;

import com.fasterxml.jackson.core.type.TypeReference;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.teamzemo.quasar.dto.ApiResponse;
import com.teamzemo.quasar.model.JdQuestion;
import com.teamzemo.quasar.model.JdSession;
import com.teamzemo.quasar.repository.JdQuestionRepository;
import com.teamzemo.quasar.repository.JdSessionRepository;
import com.teamzemo.quasar.security.UserPrincipal;
import com.teamzemo.quasar.service.GamificationService;
import com.teamzemo.quasar.service.GroqService;
import org.apache.pdfbox.Loader;
import org.apache.pdfbox.pdmodel.PDDocument;
import org.apache.pdfbox.text.PDFTextStripper;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.http.MediaType;
import org.springframework.http.ResponseEntity;
import org.springframework.security.core.annotation.AuthenticationPrincipal;
import org.springframework.web.bind.annotation.*;
import org.springframework.web.multipart.MultipartFile;

import java.io.IOException;
import java.util.*;
import java.util.stream.Collectors;

@RestController
@RequestMapping("/api")
public class JdController {

    private static final Logger log = LoggerFactory.getLogger(JdController.class);

    private final JdSessionRepository  jdSessionRepository;
    private final JdQuestionRepository jdQuestionRepository;
    private final GroqService            groqService;
    private final GamificationService    gamificationService;
    private final ObjectMapper           objectMapper = new ObjectMapper();

    private static final String SYSTEM_PROMPT = """
            You are an expert technical recruiter. Analyse the provided job description and return ONLY valid JSON (no markdown, no code fences) matching this exact schema:
            {
              "role": string,
              "seniority": "junior" | "mid" | "senior" | "staff" | "principal",
              "domain": string,
              "requiredSkills": string[],
              "niceToHaveSkills": string[],
              "culturalSignals": string[],
              "generatedQuestions": [
                {
                  "question": string,
                  "category": "behavioural" | "technical" | "system-design" | "culture-fit",
                  "difficulty": 1 | 2 | 3,
                  "targetSkill": string,
                  "weight": number
                }
              ]
            }
            Generate exactly 20 questions, weighted by importance to the role. Weight values should be between 0.0 and 1.0, higher = more important.""";

    public JdController(JdSessionRepository jdSessionRepository,
                        JdQuestionRepository jdQuestionRepository,
                        GroqService groqService,
                        GamificationService gamificationService) {
        this.jdSessionRepository  = jdSessionRepository;
        this.jdQuestionRepository = jdQuestionRepository;
        this.groqService           = groqService;
        this.gamificationService   = gamificationService;
    }

    /** POST /api/jd/parse  (multipart PDF or JSON body) */
    @PostMapping(value = "/jd/parse", consumes = {MediaType.MULTIPART_FORM_DATA_VALUE,
                                                    MediaType.APPLICATION_JSON_VALUE,
                                                    MediaType.APPLICATION_OCTET_STREAM_VALUE,
                                                    "*/*"})
    public ResponseEntity<ApiResponse<Map<String, Object>>> parseJD(
            @AuthenticationPrincipal UserPrincipal principal,
            @RequestParam(value = "jd", required = false) MultipartFile jdFile,
            @RequestParam(value = "jobDescription", required = false) String jobDescriptionParam,
            @RequestBody(required = false) Map<String, Object> jsonBody) {

        try {
            String userId = principal != null ? principal.getId() : null;
            String jobDescription = "";

            if (jdFile != null && !jdFile.isEmpty()) {
                try {
                    jobDescription = extractTextFromPDF(jdFile.getBytes());
                } catch (Exception e) {
                    return ResponseEntity.unprocessableEntity()
                            .body(ApiResponse.fail("Could not extract text from JD PDF: " + e.getMessage()));
                }
            } else if (jobDescriptionParam != null) {
                jobDescription = jobDescriptionParam.trim();
            } else if (jsonBody != null && jsonBody.get("jobDescription") != null) {
                jobDescription = String.valueOf(jsonBody.get("jobDescription")).trim();
            }

            if (jobDescription.length() < 50) {
                return ResponseEntity.badRequest()
                        .body(ApiResponse.fail("Job description must be at least 50 characters"));
            }

            String responseText = groqService.chatCompletion(
                    SYSTEM_PROMPT,
                    "Job Description:\n" + jobDescription,
                    "llama-3.3-70b-versatile", 0.4, 4096);

            String clean = stripMarkdownFences(responseText);
            Map<String, Object> parsedData = objectMapper.readValue(clean,
                    new TypeReference<Map<String, Object>>() {});

            JdSession jdSession = new JdSession();
            jdSession.setUserId(userId);
            jdSession.setJobDescription(jobDescription);
            jdSession.setParsedData(parsedData);
            jdSession = jdSessionRepository.save(jdSession);

            @SuppressWarnings("unchecked")
            List<Map<String, Object>> generatedQs =
                    (List<Map<String, Object>>) parsedData.get("generatedQuestions");
            if (generatedQs != null) {
                final String jdSessionId = jdSession.getId();
                List<JdQuestion> questions = generatedQs.stream().map(q -> {
                    JdQuestion jq = new JdQuestion();
                    jq.setJdSessionId(jdSessionId);
                    jq.setQuestion(String.valueOf(q.get("question")));
                    jq.setCategory(String.valueOf(q.getOrDefault("category", "")));
                    Object diff = q.get("difficulty");
                    jq.setDifficulty(diff instanceof Number n ? n.intValue() : 1);
                    jq.setTargetSkill(String.valueOf(q.getOrDefault("targetSkill", "")));
                    Object w = q.get("weight");
                    jq.setWeight(w instanceof Number n ? n.doubleValue() : 0.5);
                    return jq;
                }).collect(Collectors.toList());
                jdQuestionRepository.saveAll(questions);
            }

            if (userId != null) {
                new Thread(() -> gamificationService.onJDParsed(userId)).start();
            }

            Map<String, Object> data = new LinkedHashMap<>();
            data.put("jdSessionId",      jdSession.getId());
            data.put("role",             parsedData.get("role"));
            data.put("seniority",        parsedData.get("seniority"));
            data.put("domain",           parsedData.get("domain"));
            data.put("requiredSkills",   parsedData.get("requiredSkills"));
            data.put("niceToHaveSkills", parsedData.get("niceToHaveSkills"));
            data.put("culturalSignals",  parsedData.get("culturalSignals"));
            data.put("questions",        generatedQs);

            return ResponseEntity.ok(ApiResponse.ok("Job description parsed successfully", data));

        } catch (Exception e) {
            log.error("Parse JD error", e);
            return ResponseEntity.internalServerError().body(ApiResponse.fail("Failed to parse job description"));
        }
    }

    /** GET /api/jd/:jdSessionId/questions */
    @GetMapping("/jd/{jdSessionId}/questions")
    public ResponseEntity<ApiResponse<List<JdQuestion>>> getJDQuestions(@PathVariable String jdSessionId) {
        try {
            List<JdQuestion> questions = jdQuestionRepository
                    .findByJdSessionIdOrderByWeightDesc(jdSessionId);
            return ResponseEntity.ok(ApiResponse.ok("Questions retrieved", questions));
        } catch (Exception e) {
            log.error("Get JD questions error", e);
            return ResponseEntity.internalServerError().body(ApiResponse.fail("Failed to get questions"));
        }
    }

    // ── Helpers ───────────────────────────────────────────────────────

    public static String extractTextFromPDF(byte[] bytes) throws IOException {
        try (PDDocument doc = Loader.loadPDF(bytes)) {
            return new PDFTextStripper().getText(doc);
        }
    }

    static String stripMarkdownFences(String text) {
        text = text.trim();
        if (text.startsWith("```")) {
            text = text.replaceFirst("^```(?:json)?\\s*", "").replaceFirst("```\\s*$", "").trim();
        }
        return text;
    }
}
